### ISSUE TITLE
Clearing state failed for grain which hasn't been yet saved to AzureTable storage

### ISSUE THREAD

**Issue by SlavaSmorodinov** [dotnet/orleans#5219]
Good day to all.

Calling ClearStateAsync failed with exception when grain instance doesn't have table record.

This is because of grainState.Etag is null during grain Activation if it doesn't have a table record.
Possible solutions:
- Skip deletion for null Etag
- Using '*' as Etag value for null Etag

Exception details:

Exc level 0: System.ArgumentException: Delete requires an ETag (which may be the '*' wildcard).
   at Microsoft.WindowsAzure.Storage.Table.TableOperation.Delete(ITableEntity entity) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\Common\Table\TableOperation.Common.cs:line 190
   at Orleans.Persistence.AzureStorage.AzureTableDataManager`1.DeleteTableEntryAsync(T data, String eTag) in D:\build\agent\_work\24\s\src\Azure\Shared\Storage\AzureTableDataManager.cs:line 320
   at Orleans.Storage.AzureTableGrainStorage.GrainStateTableDataManager.Delete(GrainStateRecord record) in D:\build\agent\_work\24\s\src\Azure\Orleans.Persistence.AzureStorage\Providers\Storage\AzureTableStorage.cs:line 454
   at Orleans.Storage.AzureTableGrainStorage.DoOptimisticUpdate(Func`1 updateOperation, String grainType, GrainReference grainReference, String tableName, String currentETag) in D:\build\agent\_work\24\s\src\Azure\Orleans.Persistence.AzureStorage\Providers\Storage\AzureTableStorage.cs:line 160
   at Orleans.Storage.AzureTableGrainStorage.ClearStateAsync(String grainType, GrainReference grainReference, IGrainState grainState) in D:\build\agent\_work\24\s\src\Azure\Orleans.Persistence.AzureStorage\Providers\Storage\AzureTableStorage.cs:line 139
   at Orleans.Core.StateStorageBridge`1.ClearStateAsync() in D:\build\agent\_work\24\s\src\Orleans.Runtime\Storage\StateStorageBridge.cs:line 130
System.ArgumentException: Delete requires an ETag (which may be the '*' wildcard).
   at Microsoft.WindowsAzure.Storage.Table.TableOperation.Delete(ITableEntity entity) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\Common\Table\TableOperation.Common.cs:line 190
   at Orleans.Persistence.AzureStorage.AzureTableDataManager`1.DeleteTableEntryAsync(T data, String eTag) in D:\build\agent\_work\24\s\src\Azure\Shared\Storage\AzureTableDataManager.cs:line 320
   at Orleans.Storage.AzureTableGrainStorage.GrainStateTableDataManager.Delete(GrainStateRecord record) in D:\build\agent\_work\24\s\src\Azure\Orleans.Persistence.AzureStorage\Providers\Storage\AzureTableStorage.cs:line 454
   at Orleans.Storage.AzureTableGrainStorage.DoOptimisticUpdate(Func`1 updateOperation, String grainType, GrainReference grainReference, String tableName, String currentETag) in D:\build\agent\_work\24\s\src\Azure\Orleans.Persistence.AzureStorage\Providers\Storage\AzureTableStorage.cs:line 160
   at Orleans.Storage.AzureTableGrainStorage.ClearStateAsync(String grainType, GrainReference grainReference, IGrainState grainState) in D:\build\agent\_work\24\s\src\Azure\Orleans.Persistence.AzureStorage\Providers\Storage\AzureTableStorage.cs:line 139
   at Orleans.Core.StateStorageBridge`1.ClearStateAsync() in D:\build\agent\_work\24\s\src\Orleans.Runtime\Storage\StateStorageBridge.cs:line 130

**Comment by ReubenBond** [comment]
Hi @SlavaSmorodinov, thanks for reporting.

Of the two possible solutions you presented, I'm in favor of the first (no-op deletion for `null` ETag), erring towards safety. Using '*' as the ETag when the ETag is `null` could violate consistency in some rare circumstances.

Calling `Clear` on non-existent state seems uncommon, but if you'd like to submit a PR for this then it could be considered for the next release (after 2.2.0).

**Comment by ilyalukyanov** [comment]
@ReubenBond I created a PR for this https://github.com/dotnet/orleans/pull/5577

Also hit that exception in the following scenario. 

In order to prevent Storage Table from filling up with dead entries, we have `DeleteStateOnClear=true` and some logic in `OnDeactivateAsync` that does "if this grain marked itself for deletion, then call `ClearStateAsync`". Sometimes the grain can be re-activated after that (e.g. by some retry logic), which can result in another call to `ClearStateAsync` without writing any state.

Obviously, that's easy to workaround with grain's logic (which we did), but that seems like an implementation detail, which should be hidden in the storage provider's implementation.
